src/shared/fileReadUtils.ts

```diff
@@ -0,0 +1,269 @@
+	// Determine wasRangeRead *after* validation and *before* it's used for logic
+	const wasRangeRead = requestedStartLine !== undefined || requestedEndLine !== undefined
+	const startLine0Based = requestedStartLine ? requestedStartLine - 1 : 0
+	const endLine0Based = requestedEndLine ? requestedEndLine - 1 : undefined
+
+	// Initialize result object
+	const result: ProcessedFileReadResult = {
+		relativePath,
+		totalLinesInFile,
+		isBinary,
+		wasTruncated: false,
+		wasRangeRead,
+	}
+
+	// Handle binary files
+	if (isBinary) {
+		result.notice = "File is binary. Content display may be limited."
+		return result
+	}
+
+	// Determine read strategy
+	if (wasRangeRead) {
+		// Range read logic
+		const linesArray = await readLines(absolutePath, endLine0Based, startLine0Based)
+		result.contentWithLineNumbers = addLineNumbers(linesArray, requestedStartLine || 1)
+		result.actualStartLine = requestedStartLine || 1
+		result.actualEndLine = result.actualStartLine + linesArray.length - 1
+
+		if (requestedEndLine && result.actualEndLine < requestedEndLine) {
+			result.notice = `File ended at line ${result.actualEndLine} (requested to ${requestedEndLine})`
+		}
+	} else {
+		// Full or partial read logic
+		const wasTruncated = maxReadFileLine >= 0 && totalLinesInFile > maxReadFileLine
+		result.wasTruncated = wasTruncated
+
+		if (maxReadFileLine === 0) {
+			result.notice = "Content omitted (maxReadFileLine: 0). Showing definitions if available."
+		} else if (wasTruncated) {
+			const linesArray = await readLines(absolutePath, maxReadFileLine - 1, 0)
+			result.contentWithLineNumbers = addLineNumbers(linesArray, 1)
+			result.actualStartLine = 1
+			result.actualEndLine = maxReadFileLine
+			result.notice = `Showing only ${maxReadFileLine} of ${totalLinesInFile} total lines. Use start_line and end_line if you need to read more.`
+		} else {
+			result.contentWithLineNumbers = await extractTextFromFile(absolutePath)
+			result.actualStartLine = 1
+			result.actualEndLine = totalLinesInFile
+		}
+
+		// Get source code definitions if needed
+		if ((wasTruncated && maxReadFileLine > 0) || (maxReadFileLine === 0 && !wasRangeRead)) {
+			try {
+				const definitions = await parseSourceCodeDefinitionsForFile(absolutePath, rooIgnoreController)
+				if (definitions) {
+					result.sourceCodeDefinitions = definitions
+				}
+			} catch (error) {
+				if (error instanceof Error && !error.message.startsWith("Unsupported language:")) {
+					console.error(`Error parsing definitions: ${error.message}`)
+				}
+			}
+		}
+	}
+
+	// Handle empty files
+	if (totalLinesInFile === 0 && !isBinary) {
+		result.notice = "File is empty."
+	}
+
+	return result
+}
+
+/**
+ * Formats a processed file result to a string representation
+ * @param result The processed file result
+ * @returns Formatted string
+ */
+export function formatProcessedFileResultToString(result: ProcessedFileReadResult): string {
```

Comment:
This should probably stay in `readFileTool.ts` and export it if you need it somewhere else. 

 someday functions like this will be in a tool base class to build xml without string concatenation with a proper xml builder library .